### DISCUSSION TITLE
Getting MPD to use RT priority 40.

### DISCUSSION THREAD

**Discussion by jer1956** [MusicPlayerDaemon/MPD#1455] [General]
I posted that there was a bug  in the service that meant MPD never goes real time even with a RT kernel. It was closed as they refused to accept the service was the cause of the problem..
To check for real time use
ps -eo  pid,pri_bar,ni,comm. 
MPD would have a negative pri_bar if it was real time.
There are only two ways to get real time. 
A complicated way using chrt after working out the pid.
Or a simple way. Change the service.
Add 
CPUSchedulingPolicy=fifo
CPUSchedulingPriority=40
After you reboot , when you run the PS command again you get a pri_bar of -40. You are running MPD as a real time task. 
So how many believing they are using RT are if they check?

**Comment by MaxKellermann** [comment]
As I already said in your bug report at https://github.com/MusicPlayerDaemon/MPD/issues/1450#issuecomment-1046240536 : adding `CPUSchedularProfile=fifo` as you suggested is a very bad idea. Never do that!
Instead of repeating your bad advice, please read the documentation: https://mpd.readthedocs.io/en/stable/user.html#real-time-scheduling

**Reply by jer1956** [reply]
It’s isn’t working. Becuase the default cpu profile stopped being fifo ages ago.
And in other posts where users say they can’t get real time your answer is….it’s of no benefit….dont worry.
If you won’t help users must help each other

**Reply by MaxKellermann** [reply]
It works. On my computer:
```
# ps H -q `pidof -s mpd` -o 'pid,tid,cls,rtprio,comm'
    PID     TID CLS RTPRIO COMMAND
1506117 1506117  TS      - mpd
1506117 1506120  TS      - io
1506117 1506121  FF     40 rtio
1506117 1506122  TS      - player
1506117 1506123  TS      - decoder:flac
1506117 1506124  FF     40 output:alsa
1506117 1506125  TS      - iou-wrk-1506120
```

**Reply by MaxKellermann** [reply]
> If you won’t help users must help each other

Sigh. You are not helping anybody. Your advice is bad, that's the opposite of "helping".

**Reply by jer1956** [reply]
It isn’t working. MPD is not using real-time according to that information. Whats the point of setting its prlimit to 40 if it’s stuck as a user task using  nice?
You are misleading users.

**Reply by leigh123linux** [reply]
> It isn’t working. MPD is not using real-time. What’s the point of setting its prlimit to 40 if it’s stuck as a user task using nice? You are misleading users.

Maybe your distro is broken, mpd works fine on Fedora.

```
[leigh@mpd ~]$ ps H -q `pidof -s mpd` -o 'pid,tid,cls,rtprio,comm'
    PID     TID CLS RTPRIO COMMAND
    794     794  TS      - mpd
    794     878  TS      - io
    794     879  FF     40 rtio
    794    5769  TS      - player
    794    5770  TS      - decoder:dsf
    794    5771  FF     40 output:D10
```

**Reply by jer1956** [reply]
The MPD realt time priority is _ , meaning it’s using nice.
Only the services are real time. 
I will try your command to check my system with its updated service. 
Your fedora command does not work with Debian 
It will take me a bit of time to find the  command differences

**Reply by leigh123linux** [reply]
> The MPD realt time priority is _ , meaning it’s using nice. Only the services are real time. I will try your command to check my system with its updated service. Your fedora command does not work with Debian It will take me a bit of time to find the command differences

Did you read the docs?

https://mpd.readthedocs.io/en/latest/user.html?highlight=realtime#realtime

**Reply by MaxKellermann** [reply]
> You are misleading users.

You're just one step away from being banned here for being an a** towards somebody who invests his free time to try to help you.

My theory is that you didn't read the documentation, and didn't check that your system is properly set up to allow unprivileged real-time scheduling.

**Comment by jer1956** [comment]
if i get the pid seperately your command works
ps H -q 513 -o pid,tid,cls,rtprio,comm
    PID     TID CLS RTPRIO COMMAND
    513     513  FF     40 mpd
    513     553  FF     40 io
    513     554  FF     40 rtio
    513    3541  FF     40 player
    513    3542  FF     40 decoder:flac
    513    3543  FF     40 iou-wrk-553
    513    3544  FF     40 output:My ALSA 
    513   11951  TS      - output:My ALSA 

Now that’s what I call running MPD in real time.
 And why do you keep editing out posts you don’t like in a discussion?

**Reply by MaxKellermann** [reply]
Running all threads with real-time priority is a serious misconfiguration that will lead to trouble. That cannot be pointed out enough. Please stop spreading your misinformation.

**Reply by leigh123linux** [reply]
> if i get the pid seperately your command works ps H -q 513 -o pid,tid,cls,rtprio,comm PID TID CLS RTPRIO COMMAND 513 513 FF 40 mpd 513 553 FF 40 io 513 554 FF 40 rtio 513 3541 FF 40 player 513 3542 FF 40 decoder:flac 513 3543 FF 40 iou-wrk-553 513 3544 FF 40 output:My ALSA 513 11951 TS - output:My ALSA
> 
> Now that’s what I call running MPD in real time.

So you want the mpd database and everything else not sound related running high priority and possibly degrading sound?

**Reply by jer1956** [reply]
Then you should stop saying you have a real time setup. 
You don’t! 
That’s the real  misinformation.

**Reply by MaxKellermann** [reply]
> Then you should stop saying you have a real time setup. You don’t!

WTF ... I have a feeling you have no idea what you are talking about. You are making a clown of yourself.

**Reply by jer1956** [reply]
No….you are an egotist who can’t stand critism.

**Reply by MaxKellermann** [reply]
Enough is enough. I have just banned @jer1956.

**Comment by twojstaryzdomu** [comment]
@MaxKellermann Not to rescind an old discussion, just to ask the validity of my setup. I followed your guide but it wasn't enough when the system was under load and additional steps had to be taken to improve real-time audio.

I understand the problem when everything runs real-time but what about `RR` class and `PTPRIO=50`?

```
# cat /etc/systemd/system/mpd.service.d/10_scheduling.conf 
[Service]
CPUSchedulingPolicy=rr
CPUSchedulingPriority=50
# ps H -q `pidof -s mpd` -o 'pid,tid,cls,rtprio,comm'
  PID   TID CLS RTPRIO COMMAND
20923 20923  RR     50 mpd
20923 20929  RR     50 io
20923 20930  FF     40 rtio
20923 20931  RR     50 player
20923 20932  RR     50 decoder:flac
20923 20952  FF     40 output:Pulseaud
20923 20953  TS      - threaded-ml
```

It works in combination with `pulseaudio` (spawned system-wide to secure shared memory but as user mpd), some additional details below. Note `RTPRIO=51`, one notch higher than mpd.
```
# cat /etc/systemd/system/pulseaudio.service.d/10_scheduling.conf 
[Service]
ExecStartPost=chrt -p 51 ${MAINPID}
IOSchedulingPriority=1
IOSchedulingClass=realtime
# ps H -q `pidof -s pulseaudio` -o 'pid,tid,cls,rtprio,comm'
  PID   TID CLS RTPRIO COMMAND
  919   919  RR     51 pulseaudio
  919   927  RR     99 alsa-sink-USB A
# grep Limit /etc/systemd/system/pulseaudio.service
LimitRTPRIO=99
LimitRTTIME=infinity
LimitNICE=-20
LimitMEMLOCK=infinity
# cat /etc/pulse/daemon.conf
cpu-limit = no
high-priority = yes
nice-level = -12
rlimit-nice = -12
realtime-scheduling = yes
realtime-priority = 99
rlimit-rtprio = 99
exit-idle-time = -1
default-sample-format = s24le
default-sample-rate = 96000
resample-method = soxr-mq
enable-shm = yes
```
This setup goes back a few years, `IOSchedulingPriority` as I recall helped resolve skipping when under heavy load. The environment with the above setup has meagre resources.
